File: week-02-2D-1D-models/Codes/MorrisLecar.m
function dudt = MorrisLecar(t,u,p)
  % function dudt = MorrisLecar(t,u,p)
  % Computes the vectorfield of the Morris-Lecar model
  % The notation follows ET, Section 3.7.
  % Input varaibles: 
  %      t: time, expressed as a scalar
  %      u: the vector variables u = [v; w]
  %      p: the control parameters p = [a epsi gamma I]
  %      dudt: the vectorfield, with components ordered as in u

  %% Unpack parameters
  Cm   = p(1); 
  gCa  = p(2); 
  gK   = p(3); 
  gL   = p(4); 
  eCa  = p(5); 
  eK   = p(6); 
  eL   = p(7); 
  phi  = p(8); 
  V1   = p(9);
  V2   = p(10);
  V3   = p(11);
  V4   = p(12);
  IApp = p(13);

  %% Unpack state variables
  v = u(1);
  n = u(2);

  %% Activation and inactivation variables
  mInf = 0.5*( 1 + tanh((v-V1)/V2)); 
  nInf = 0.5*( 1 + tanh((v-V3)/V4)); 
  tau  = 1./( cosh((v-V3)/(2*V4)) );

  %% Currents
  ICa  = gCa*mInf*(v-eCa);
  IK   = gK*n*(v-eK);
  IL   = gL*(v-eL);

  %% Assemble the vectorfield 
  dudt = zeros(size(u));
  dudt(1) = (-ICa - IK - IL + IApp)/Cm;
  dudt(2) = phi/tau * (nInf - n);

end

File: single-neuronal-models/py2/MorrisLecar.py
from numpy import ndarray, array, tanh, cosh

# 2D Dynamical System
def MorrisLecar(
        t:float, # time
        # these are the (state determining) variables that are solved for
        variables:list[float],
        # function that determines the applied current I based on the timepoint (t)
        Ifunc:callable,
        # these params determine the behavior of the model
        params:list[float] = [20, 2, -60, 8, -84, 4, 120, -1.2, 18, 12, 17.4]
        ) -> ndarray: # returns du/dt with u containing state variables u 
    
    # TODO: MorrisLecar model logic
    # voltage, recovery variable
    v, n = variables

    Cm, gL, eL, gK, eK, gCa, eCa, v1, v2, v3, v4 = params

    # v1 & v3 control location of curve (left-right), v2 & v4 control steepness / width / shape
    minf = (1+tanh((v-v1)/v2))/2                         
    ninf = (1+tanh((v-v3)/v4))/2              
    tau = 1/cosh((v-v3)/(2*v4))                         

    # Currents:
    # Leak
    IL = - gL*(v - eL)
    # recovery (potassium)
    IK = - gK*n*(v - eK)
    # active (sodium / calcium in morris lecar)
    ICa = - gCa*minf*(v - eCa)
    # applied current
    IApp = Ifunc(t)

    dvdt = (IApp + IL + IK + ICa) / Cm
    dndt = (ninf - n) / tau

    # 2D Dynamical System
    return array([dvdt,
                  dndt])